### PR TITLE
Add HasPhoto DAV plugin

### PR DESCRIPTION
**Pull request by skjnldsv**
This will allow us to know if the contact have a PHOTO set without having to retrieve the FULL vcard and download extra data.

### CODE REVIEW

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

File: apps/dav/lib/CardDAV/Xml/HasPhoto.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+namespace OCA\DAV\CardDAV\Xml;
+
+use Sabre\VObject\Component\VCard;
+use Sabre\VObject\Reader;
+use Sabre\Xml\Writer;
+use Sabre\Xml\XmlSerializable;
+
+class HasPhoto implements XmlSerializable {
+	const NS_Nextcloud = 'http://nextcloud.com/ns';
```
Comment (MorrisJobke) [inline apps/dav/lib/CardDAV/Xml/HasPhoto.php]:
```suggestion
	const NS_NEXTCLOUD = 'http://nextcloud.com/ns';
```

**Comment by skjnldsv** [comment]
@MorrisJobke if you want to test, just open the contacts app and copy the curl link of the vcf request to the contact (the one with an UUID in the requests list)
Paste it on your terminal and add the `<x4:has-photo xmlns:x4="http://nextcloud.com/ns"/>` into the `<x0:prop>` xml data :)

File: apps/dav/lib/CardDAV/Xml/HasPhoto.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	private $cardData;
+
+	/**
+	 * @param string $cardData
+	 */
+	public function __construct($cardData) {
+		$this->cardData = $cardData;
+	}
+
+	function xmlSerialize(Writer $writer) {
```
Comment (rullzer) [inline apps/dav/lib/CardDAV/Xml/HasPhoto.php]: If I'm not mistaken there is no need to do this xmlSedrializer stuff. You could just have  a method that returns true or false... 

Or is there a reason you want to write it yourself :wink:

Comment (skjnldsv) [reply apps/dav/lib/CardDAV/Xml/HasPhoto.php]: I have no idea! This whole dav plugin is a bit weird for me :see_no_evil: 
Can you suggest an edit?

**Comment by rullzer** [comment]
@skjnldsv pushed a fix. This should be good enough I think.

**Comment by rullzer** [comment]
> @skjnldsv pushed a fix. This should be good enough I think.

If not just drop my commit :wink:

**Comment by skjnldsv** [comment]
Thanks!!
Somehow I get 404 on every `/remote.php/dav/addressbooks/admin/` (try loading the contacts app) on my setup now. Can anyone reproduce?

**Comment by skjnldsv** [comment]
Well, I couldn't reproduce the issue! @rullzer all works fine!
Rebased and ready to be merged! :)

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Comment by skjnldsv** [comment]
Failures unrelated: timeouts or `sh: 1: kill: No such process`

**Review by rullzer — APPROVED** [review]

**Comment by georgehrke** [comment]
Hm :/

Ideally this would have been implemented upstream in Sabre/DAV.
CardDAV specifies support for a `no value` attribute on a property:
https://tools.ietf.org/html/rfc6352#section-10.4.2

This way, the server would just return an empty `PHOTO` property.

(Sabre/DAV doesn't support it yet, but we could have implemented it there.)

**Comment by skjnldsv** [comment]
@georgehrke agreed, but then we'll revert this!
My php knowledge is far beyond being able to implement this upstream in Sabre/DAV :confused:

**Pull request merged**